From 70837e0fa00942e05da5f44a9159b506e867fdd3 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Mon, 4 Feb 2019 05:42:24 -0800
Subject: [PATCH 22/31] UBUNTU: SAUCE: apparmor: add an apparmorfs entry to
 access current attrs

There are use cases where the proc filesystem is not available but
access to the apparmor attr interface for change_profile and
change_hat is desired.

Provide an apparmorfs based interface for the current task that
mirrors what is provided in proc.

Signed-off-by: John Johansen <john.johansen@canonical.com>
Signed-off-by: Seth Forshee <seth.forshee@canonical.com>
---
 security/apparmor/apparmorfs.c         | 66 ++++++++++++++++++++++++++
 security/apparmor/include/apparmorfs.h |  3 ++
 security/apparmor/lsm.c                |  6 +--
 3 files changed, 71 insertions(+), 4 deletions(-)

diff --git a/security/apparmor/apparmorfs.c b/security/apparmor/apparmorfs.c
index 2f3efac3fe88..b03b1f92f701 100644
--- a/security/apparmor/apparmorfs.c
+++ b/security/apparmor/apparmorfs.c
@@ -2196,6 +2196,63 @@ static const struct file_operations aa_sfs_profiles_fops = {
 	.release = profiles_release,
 };
 
+static ssize_t attr_read(struct file * file, char __user * buf, size_t count,
+			 loff_t *ppos)
+{
+	char *p = NULL;
+	ssize_t length;
+
+	length = apparmor_getprocattr(current,
+				      (char*)file->f_path.dentry->d_name.name,
+				      &p);
+	if (length > 0)
+		length = simple_read_from_buffer(buf, count, ppos, p, length);
+	kfree(p);
+
+	return length;
+}
+
+static ssize_t attr_write(struct file * file, const char __user * buf,
+			  size_t count, loff_t *ppos)
+{
+	void *page;
+	ssize_t length;
+
+	length = -ESRCH;
+
+	if (count > PAGE_SIZE)
+		count = PAGE_SIZE;
+
+	/* No partial writes. */
+	length = -EINVAL;
+	if (*ppos != 0)
+		goto out;
+
+	page = memdup_user(buf, count);
+	if (IS_ERR(page)) {
+		length = PTR_ERR(page);
+		goto out;
+	}
+
+	/* Guard against adverse ptrace interaction */
+	length = mutex_lock_interruptible(&current->signal->cred_guard_mutex);
+	if (length < 0)
+		goto out_free;
+
+	length = apparmor_setprocattr(file->f_path.dentry->d_name.name, page,
+				      count);
+	mutex_unlock(&current->signal->cred_guard_mutex);
+out_free:
+	kfree(page);
+out:
+	return length;
+}
+
+static const struct file_operations attr_fops = {
+	.read = attr_read,
+	.write = attr_write,
+};
+
 
 /** Base file system setup **/
 static struct aa_sfs_entry aa_sfs_entry_file[] = {
@@ -2273,6 +2330,7 @@ static struct aa_sfs_entry aa_sfs_entry_query[] = {
 	AA_SFS_DIR("label",			aa_sfs_entry_query_label),
 	{ }
 };
+
 static struct aa_sfs_entry aa_sfs_entry_features[] = {
 	AA_SFS_DIR("policy",			aa_sfs_entry_policy),
 	AA_SFS_DIR("domain",			aa_sfs_entry_domain),
@@ -2291,6 +2349,13 @@ static struct aa_sfs_entry aa_sfs_entry_features[] = {
 	{ }
 };
 
+static struct aa_sfs_entry aa_sfs_entry_attr[] = {
+	AA_SFS_FILE_FOPS("current", 0666, &attr_fops),
+	AA_SFS_FILE_FOPS("prev", 0444, &attr_fops),
+	AA_SFS_FILE_FOPS("exec", 0666, &attr_fops),
+	{ }
+};
+
 static struct aa_sfs_entry aa_sfs_entry_apparmor[] = {
 	AA_SFS_FILE_FOPS(".access", 0666, &aa_sfs_access),
 	AA_SFS_FILE_FOPS(".stacked", 0444, &seq_ns_stacked_fops),
@@ -2299,6 +2364,7 @@ static struct aa_sfs_entry aa_sfs_entry_apparmor[] = {
 	AA_SFS_FILE_FOPS(".ns_name", 0444, &seq_ns_name_fops),
 	AA_SFS_FILE_FOPS("profiles", 0444, &aa_sfs_profiles_fops),
 	AA_SFS_DIR("features", aa_sfs_entry_features),
+	AA_SFS_DIR("attr", aa_sfs_entry_attr),
 	{ }
 };
 
diff --git a/security/apparmor/include/apparmorfs.h b/security/apparmor/include/apparmorfs.h
index 6e14f6cecdb9..7eaa0003dd2d 100644
--- a/security/apparmor/include/apparmorfs.h
+++ b/security/apparmor/include/apparmorfs.h
@@ -117,4 +117,7 @@ struct aa_loaddata;
 void __aa_fs_remove_rawdata(struct aa_loaddata *rawdata);
 int __aa_fs_create_rawdata(struct aa_ns *ns, struct aa_loaddata *rawdata);
 
+int apparmor_getprocattr(struct task_struct *task, char *name, char **value);
+int apparmor_setprocattr(const char *name, void *value, size_t size);
+
 #endif /* __AA_APPARMORFS_H */
diff --git a/security/apparmor/lsm.c b/security/apparmor/lsm.c
index ed790ba4fc73..683195e12ea3 100644
--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -570,8 +570,7 @@ static int apparmor_sb_pivotroot(const struct path *old_path,
 	return error;
 }
 
-static int apparmor_getprocattr(struct task_struct *task, char *name,
-				char **value)
+int apparmor_getprocattr(struct task_struct *task, char *name, char **value)
 {
 	int error = -ENOENT;
 	/* released below */
@@ -597,8 +596,7 @@ static int apparmor_getprocattr(struct task_struct *task, char *name,
 	return error;
 }
 
-static int apparmor_setprocattr(const char *name, void *value,
-				size_t size)
+int apparmor_setprocattr(const char *name, void *value, size_t size)
 {
 	char *command, *largs = NULL, *args = value;
 	size_t arg_size;
-- 
2.25.1

